File: ts/11.binary-tree/find-closest-value.ts
import { BinaryTree } from './BinaryTree';
import { Node } from './Node';

export function findClosestNumber(bt: BinaryTree<number>, target: number) {
    let queue = [bt.getRoot()];
    let seen = new Map<number, boolean>();
    let closest = Infinity;

    while (queue.length > 0) {
        const node = queue.shift() as Node<number>;

        node.left && target < node.value && queue.push(node.left);
        node.right && target > node.value && queue.push(node.right);

        if (!seen.get(node.value)) {
            seen.set(node.value, true);
            if (Math.abs(node.value - target) < Math.abs(closest - target)) {
                closest = node.value;
            }
        }
    }

    return closest;
}

function dfs(
    node: Node<number> | null,
    target: number,
    closest: number
): number {
    if (!node) return closest;

    if (Math.abs(target - node.value) < Math.abs(target - closest)) {
        closest = node.value;
    }

    if (node.right && target > node.value) {
        return dfs(node.right, target, closest);
    } else if (node.left && target < node.value) {
        return dfs(node.left, target, closest);
    } else {
        return closest;
    }
}

export function findClosestNumberDfs(bt: BinaryTree<number>, target: number) {
    return dfs(bt.getRoot(), target, Infinity);
}

File: golang/11.binary-tree/main/find_closest_value.go
package main

import (
	"math"

	"github.com/AmirAhmadzadeh/binary_tree/binary_tree"
	"github.com/AmirAhmadzadeh/binary_tree/node"
)

func findClosestValueInBinaryTree(tree *binary_tree.BinaryTree, target int) float64 {
	result := math.Inf(+1)
	queue := []*node.Node{tree.GetRoot()}

	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]
		if math.Abs(float64(vertex.Value()-target)) <= math.Abs(result-float64(target)) {
			result = float64(vertex.Value())
		}
		if vertex.Left() != nil && target < vertex.Value() {
			queue = append(queue, vertex.Left())
		}
		if vertex.Right() != nil && target > vertex.Value() {
			queue = append(queue, vertex.Right())
		}
	}

	return result
}

func findClosestValueInBinaryTreeDfsCall(vertex *node.Node, target float64, closest float64) float64 {
	if vertex == nil {
		return closest
	}
	if math.Abs(target-float64(vertex.Value())) < math.Abs(target-closest) {
		closest = float64(vertex.Value())
	}

	if vertex.Right() != nil && target > float64(vertex.Value()) {
		return findClosestValueInBinaryTreeDfsCall(vertex.Right(), target, closest)
	} else if vertex.Left() != nil && target < float64(vertex.Value()) {
		return findClosestValueInBinaryTreeDfsCall(vertex.Left(), target, closest)
	}

	return closest
}

func findClosestValueInBinaryTreeDfs(binaryTree *binary_tree.BinaryTree, target float64) float64 {
	return findClosestValueInBinaryTreeDfsCall(binaryTree.GetRoot(), target, math.Inf(1))
}
